docs-internal/config-metadata.md

```diff
@@ -0,0 +1,404 @@
+               - provider: "provider"
+                 config: "something"
+     ```
+type - either a type directly mapped to an object (String, Integer, URI etc.), or a nested type defined either in 
+     this or in another module. Nested configuration may be in another metadata json.
+kind LIST: this is a list of values (either simple values, or objects)
+kind MAP: this is a map of values, using keys of the map as keys, and values of the map as values, `type` defines 
+     the value type (must be a simple value), key is expected to be `java.lang.String`
+kind VALUE: either a simple value (String, Long etc.) or a nested object (this is the default)
+method: method to be called to configure this option (may be a static factory method, such as `create(Config)`, or a builder method)
+merge: if set to `true` (default is `false`) this option's key is ignored and all its nested keys are inserted
+     directly into the parent option
+experimental: experimental options may change without warning even between minor releases (such as support for Loom in Java)
+required: if set to `true` (default is `false`) this option must be present in configuration if parent option is present
+provider: this option expects a type with matching `provides` - such as security providers configuration
+deprecated: this option should no longer be used, there is another option allowing the same (usually when renaming options,
+     fixing inconsistencies etc.); description should provide more details
+defaultValue: for simple values, this defines the string default value
+
+Example of a simple optional element with defa# Configuration metadata
+
+A new module `helidon-config-metadata` now exists with annotations that can be used in Helidon source code
+the document what configuration is used.
+
+These annotations are processed by `helidon-config-metadata-processor`.
+
+## Add metadata to a configurable component
+
+To add meta configuration:
+
+1. Add the following dependency to your `pom.xml`:
+```xml
+<dependency>
+    <groupId>io.helidon.config</groupId>
+    <artifactId>helidon-config-metadata</artifactId>
+    <scope>provided</scope>
+    <optional>true</optional>
+</dependency>
+```
+2. Add the following compiler plugin configuration to add the annotation processor:
+```xml
+<build>
+    <plugins>
+        <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <configuration>
+                <forceJavacCompilerUse>true</forceJavacCompilerUse>
+                <annotationProcessorPaths>
+                    <path>
+                        <groupId>io.helidon.config</groupId>
+                        <artifactId>helidon-config-metadata-processor</artifactId>
+                        <version>${helidon.version}</version>
+                    </path>
+                </annotationProcessorPaths>
+            </configuration>
+        </plugin>
+    </plugins>
+</build>
```

Comment:
This is a lot of boilerplate to get right in multiple components' `pom.xml` files, especially if it ever has to change for some reason. 

Is the annotation processor fast enough so we could set it on by default for all components? If it finds nothing configurable it would not output anything.

If speed is an issue, maybe the plug-in could first check to see if the module depends on the metadata artifact and, if so, then do its full processing?

There might be other, better techniques to accomplish this. It would just be very nice if the build did the right thing without me (as a component developer) having to insert this identical boilerplate in all affected modules.

Reply:
@tjquinno 

(not doing a quote reply as this is the very next comment)

`config/metadata` is a new config module and is used through-out the repository to annotate config related code.
One may have to update the metadata and processor when making certain kind of changes to the config API.

If you look at the processor implementation you can see it visits config code. E.g.
```java
configType = elementUtils.getTypeElement("io.helidon.config.Config").asType();
```

Reply:
Yes, I will need to dive a bit deeper into this if we can find a better way of handling it.

Reply:
I have found a way to do this. I have removed all dependencies from the annotation processor. Now it can be simply added as an optional provided dependency. 
This required creating a few classes  to write JSON output, but I think worth it to reduce dependencies...